## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [24]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [11]:
# View the column names in the data
print(df.columns)
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [37]:
df.count()

first             54639
last              54639
job               54639
phone             54639
email             54639
client_id         54639
order_id          54639
order_date        54639
order_week        54639
order_year        54639
item_id           54639
category          54639
subcategory       54639
unit_price        54639
unit_cost         54639
unit_weight       54639
qty               54639
line_number       54639
subtotal          54639
total_weight      54639
shipping_price    54639
sales_tax         54639
total_price       54639
profit            54639
dtype: int64

In [38]:
df.dtypes


first              object
last               object
job                object
phone              object
email              object
client_id           int64
order_id            int64
order_date         object
order_week          int64
order_year          int64
item_id            object
category           object
subcategory        object
unit_price        float64
unit_cost         float64
unit_weight       float64
qty                 int64
line_number         int64
subtotal          float64
total_weight      float64
shipping_price    float64
sales_tax         float64
total_price       float64
profit            float64
dtype: object

In [12]:
# Use the describe function to gather some basic statistics
basic_statistics = df.describe()
print(basic_statistics)

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [13]:
# Use this space to do any additional research
# and familiarize yourself with the data.
all_statistics = df.describe(include='all')

# Display the statistics
print(all_statistics)

              first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      o

In [14]:
# What five item categories had the most entries?
category_counts = df['category'].value_counts()
top_five_categories = category_counts.head(5)
print(top_five_categories)

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64


In [34]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
print(f"The category with the most entries is '{top_category}'")

# Filter the DataFrame to include only entries from the top category
top_category_df = df[df['category'] == top_category]
print(f"Entries for the top category '{top_category}':\n", top_category_df)

# Count the subcategories within this top category
subcategory_counts = top_category_df['subcategory'].value_counts()
print(f"Subcategory counts within the top category '{top_category}':\n", subcategory_counts)

# Get the subcategory with the most entries and its count
top_subcategory = subcategory_counts.idxmax()
top_subcategory_count = subcategory_counts.max()
print(f"top_subcategory:{top_subcategory}")
print(f"top_subcategory_count:{top_subcategory_count}")
# Display the top subcategory and its count
print(f"The subcategory with the most entries within the top category '{top_category}' is '{top_subcategory}' with {top_subcategory_count} entries.")

The category with the most entries is 'consumables'
Entries for the top category 'consumables':
            first       last                                      job  \
1        Tiffany      Myers                          Music therapist   
3         Nathan      Baker                    Accounting technician   
4      Christina   Schwartz                             Chiropractor   
6          Brian      Moore                                  Barista   
7          Jesus     Wright                       Theme park manager   
...          ...        ...                                      ...   
54629      Linda  Henderson  Scientist, research (physical sciences)   
54632     Austin     Fuller                       Investment analyst   
54633     Tamara      Jones                        Fish farm manager   
54635    Derrick      Moore                                 Musician   
54636     Monica  Gutierrez                         Graphic designer   

                     phone            

In [16]:
# Which five clients had the most entries in the data?
client_id = df['client_id'].value_counts()
top_five_clients = client_id.head(5)
print(top_five_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [36]:
client_id = df['client_id'].value_counts()
top_five_clients = client_id.head(5)
print(top_five_clients)
# Store the client ids of those top 5 clients in a list.
top_five_client_ids = top_five_clients.index.tolist()
print(top_five_client_ids)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64
[33615, 66037, 46820, 38378, 24741]


In [18]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()
top_client_df = df[df['client_id'] == top_client]
total_units_ordered = top_client_df['qty'].sum()
print(f"The client with the most entries is '{top_client}' who ordered a total of {total_units_ordered} units.")



The client with the most entries is '33615' who ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [3]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
import pandas as pd
df = pd.DataFrame(df)
subtotal = df['unit_price'] * df['qty']   
print(subtotal)
#df['subtotal'] = df['unit_price'] * df['qty']
#print(df['subtotal'].sum())



0        115164.00
1           523.95
2           527.28
3          1056.18
4          3902.00
           ...    
54634      2743.29
54635      9709.73
54636     31188.50
54637       165.76
54638      3103.31
Length: 54639, dtype: float64


In [4]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')
df.head()
#Calculate total weight for each order line
df['total_weight'] = df['unit_weight'] * df['qty']

# Define the shipping price based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)
print(df[['order_id', 'total_weight', 'shipping_price']])




       order_id  total_weight  shipping_price
0       8953482        787.50         5512.50
1       8069089         31.29          312.90
2       1902144         65.52          458.64
3       9031802         35.67          356.70
4       1322274        928.60         6500.20
...         ...           ...             ...
54634   9021716         74.25          519.75
54635   6290153        549.90         3849.30
54636   8692622       1976.00        13832.00
54637   7592730       2020.48        14143.36
54638   7489403       2134.17        14939.19

[54639 rows x 3 columns]


In [21]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')
df.head()
# Calculate subtotal for each order line
df['subtotal'] = df['unit_price'] * df['qty']

# Calculate total weight for each order line
df['total_weight'] = df['unit_weight'] * df['qty']

# Define the shipping price based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)

# Calculate the sales tax
df['sales_tax'] = df['subtotal'] * 0.0925

# Calculate the total price
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']

print(df[['order_id', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax', 'total_price']])


       order_id   subtotal  total_weight  shipping_price     sales_tax  \
0       8953482  115164.00        787.50         5512.50  10652.670000   
1       8069089     523.95         31.29          312.90     48.465375   
2       1902144     527.28         65.52          458.64     48.773400   
3       9031802    1056.18         35.67          356.70     97.696650   
4       1322274    3902.00        928.60         6500.20    360.935000   
...         ...        ...           ...             ...           ...   
54634   9021716    2743.29         74.25          519.75    253.754325   
54635   6290153    9709.73        549.90         3849.30    898.150025   
54636   8692622   31188.50       1976.00        13832.00   2884.936250   
54637   7592730     165.76       2020.48        14143.36     15.332800   
54638   7489403    3103.31       2134.17        14939.19    287.056175   

         total_price  
0      131329.170000  
1         885.315375  
2        1034.693400  
3        1510.57665

In [22]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

# Calculate subtotal for each order line
df['subtotal'] = df['unit_price'] * df['qty']

# Calculate total weight for each order line
df['total_weight'] = df['unit_weight'] * df['qty']

# Define the shipping price based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)

# Calculate the sales tax
df['sales_tax'] = df['subtotal'] * 0.0925

# Calculate the total price
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']

# Calculate the cost of each line
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

print(df[['order_id', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax', 'total_price', 'line_cost']])

# Convert total_price to integers to remove decimals
df['total_price'] = df['total_price'].astype(int)

# Print the resulting DataFrame with the selected columns
print(df[['order_id', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax', 'total_price']])







       order_id   subtotal  total_weight  shipping_price     sales_tax  \
0       8953482  115164.00        787.50         5512.50  10652.670000   
1       8069089     523.95         31.29          312.90     48.465375   
2       1902144     527.28         65.52          458.64     48.773400   
3       9031802    1056.18         35.67          356.70     97.696650   
4       1322274    3902.00        928.60         6500.20    360.935000   
...         ...        ...           ...             ...           ...   
54634   9021716    2743.29         74.25          519.75    253.754325   
54635   6290153    9709.73        549.90         3849.30    898.150025   
54636   8692622   31188.50       1976.00        13832.00   2884.936250   
54637   7592730     165.76       2020.48        14143.36     15.332800   
54638   7489403    3103.31       2134.17        14939.19    287.056175   

         total_price  line_cost  
0      131329.170000   85597.05  
1         885.315375     629.79  
2        

In [23]:
# Create a column for the profit of each line using line cost and line price
#Calculate the profit for each line
df['profit'] = df['total_price'] - df['line_cost']

# Sort the DataFrame by total_price in descending order and select the top 10
top_orders = df.sort_values(by='total_price', ascending=False).head(10)

print(top_orders[['order_id', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax', 'total_price', 'line_cost', 'profit']])




       order_id      subtotal  total_weight  shipping_price     sales_tax  \
21801   3833568  5.844743e+08     356241.96      2493693.72  5.406387e+07   
15112   7263857  2.061214e+08    1408472.78      9859309.46  1.906623e+07   
25972   6057452  9.810295e+07    9009055.62     63063389.34  9.074523e+06   
23929   4594154  1.482472e+08     494201.25      3459408.75  1.371287e+07   
12597   3989744  4.873314e+07    5835503.73     40848526.11  4.507815e+06   
11740   7071092  8.245314e+07     499148.80      3494041.60  7.626916e+06   
35079   6656709  7.371976e+07      18986.48       132905.36  6.819077e+06   
50048   3089707  4.018383e+07    3284754.88     22993284.16  3.717004e+06   
4063    5167984  5.635518e+07     227228.76      1590601.32  5.212854e+06   
23123   9226575  4.596323e+07     969731.50      6788120.50  4.251599e+06   

       total_price     line_cost        profit  
21801    641031876  3.987535e+08  2.422784e+08  
15112    235046925  1.270239e+08  1.080231e+08  
25972

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
###EXTRA STUFF JUST FOR ME#########
# Convert total_price to integers to remove decimals
#df['total_price'] = df['total_price'].astype(int)

# Calculate the cost of each line
#df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

# Calculate the profit for each line
#df['profit'] = df['total_price'] - df['line_cost']


# Print the resulting DataFrame with the selected columns
#print(df[['order_id', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax', 'total_price', 'line_cost', 'profit']])

# Sort the DataFrame by total_price in descending order and select the top 10
#top_orders = df.sort_values(by='total_price', ascending=False).head(10)
#print(top_orders[['order_id', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax', 'total_price', 'line_cost', 'profit']])


In [ ]:
import pandas as pd

# Reading the CSV file
df = pd.read_csv('Resources/client_dataset.csv')

# Filter for the specific order ID
df = df[df['order_id'] == 2173913]

# Calculate subtotal for each order line
df['subtotal'] = df['unit_price'] * df['qty']
subtotal_sum = df['subtotal'].sum()
print("Subtotal sum:", round(subtotal_sum, 2))

# Calculate total weight for each order line
df['total_weight'] = df['unit_weight'] * df['qty']
total_weight_sum = df['total_weight'].sum()
print("Total weight sum:", round(total_weight_sum, 2))

# Define the shipping price based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)
shipping_price_sum = df['shipping_price'].sum()
print("Shipping price sum:", round(shipping_price_sum, 2))

# Calculate the sales tax
df['sales_tax'] = df['subtotal'] * 0.0925
sales_tax_sum = df['sales_tax'].sum()
print("Sales tax sum:", round(sales_tax_sum, 2))

# Calculate the total price
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']
total_price_sum = df['total_price'].sum()
print("Total price sum:", round(total_price_sum, 2))

# Verify the total price for the specific order ID
order_id_to_check = 2173913
expected_total_price = 162388.71

# Verify the total price for the specific order ID
order = df[df['order_id'] == order_id_to_check]

if order.empty:
    print(f"Order ID {order_id_to_check} not found.")
else:
    actual_total_price = order['total_price'].sum()
    print(f"Order ID {order_id_to_check} actual total price: {round(actual_total_price, 2)}")
    if abs(actual_total_price - expected_total_price) < 1e-2:  # Allowing for floating-point precision issues
        print(f"Order ID {order_id_to_check} has the expected total price of ${expected_total_price}")
    else:
        print(f"Order ID {order_id_to_check} does not match the expected total price. Actual: ${round(actual_total_price, 2)}, Expected: ${expected_total_price}")

In [25]:
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')
df = df[df['order_id'] == 2173913]
df['subtotal'] = df['unit_price'] * df['qty']
print("Subtotal sum:", df['subtotal'].sum())
df['total_weight'] = df['unit_weight'] * df['qty']
print("Total weight sum:", df['total_weight'].sum())
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)
print("Shipping price sum:", df['shipping_price'].sum())
df['sales_tax'] = df['subtotal'] * 0.0925
print("Sales tax sum:", df['sales_tax'].sum())
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']
print("Total price sum:", df['total_price'].sum())
order_id_to_check = 2173913
expected_total_price = 162388.71

order = df[df['order_id'] == order_id_to_check]

if order.empty:
    print(f"Order ID {order_id_to_check} not found.")
else:
    actual_total_price = order['total_price'].sum()
    print(f"Order ID {order_id_to_check} actual total price: {actual_total_price}")
    if abs(actual_total_price - expected_total_price) < 1e-2:
        print(f"Order ID {order_id_to_check} has the expected total price of ${expected_total_price}")
    else:
        print(f"Order ID {order_id_to_check} does not match the expected total price. Actual: ${actual_total_price}, Expected: ${expected_total_price}")


Subtotal sum: 109440.76
Total weight sum: 5576.57
Shipping price sum: 39198.799999999996
Sales tax sum: 10123.2703
Total price sum: 158762.83029999997
Order ID 2173913 actual total price: 158762.83029999997
Order ID 2173913 does not match the expected total price. Actual: $158762.83029999997, Expected: $162388.71


In [23]:
# Check your work using the totals above
# Check if order ID 2173913 had a total price of $162,388.71


# Check if order ID 2173913 had a total price of $162,388.71
import pandas as pd
df = pd.read_csv('Resources/client_dataset.csv')

#delete this later
df = df[df['order_id'] == 2173913]

# Calculate subtotal for each order line
df['subtotal'] = df['unit_price'] * df['qty']
print("this is the subtotal sum", df['subtotal'].sum())

# Calculate total weight for each order line
df['total_weight'] = df['unit_weight'] * df['qty']
print("this is the total_weight sum",df['total_weight'].sum())

# Define the shipping price based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7*x if x > 50 else 10*x)
print("this is the shipping_price",df['shipping_price'].sum())

# Calculate the sales tax
df['sales_tax'] = df['subtotal'] * 0.0925
print("this is the sales_tax sum", df['sales_tax'].sum())

# Calculate the total price
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax'] 
print(df['total_price'].sum())
#print(df[['total_price','subtotal','total_weight']])

# Check if order ID 2173913 had a total price of $162,388.71
order_id_to_check = 2173913
expected_total_price = 162388.71

# Verify the total price for the specific order ID
order = df[df['order_id'] == order_id_to_check]

if order.empty:
    print(f"Order ID {order_id_to_check} not found.")
else:
    actual_total_price = order['total_price'].sum()
    print(f"Order ID {order_id_to_check} actual total price: {actual_total_price}")
    if abs(actual_total_price - expected_total_price) < 1e-2:  # Allowing for floating-point precision issues
        print(f"Order ID {order_id_to_check} has the expected total price of ${expected_total_price}")
    else:
        print(f"Order ID {order_id_to_check} does not match the expected total price. Actual: ${actual_total_price}, Expected: ${expected_total_price}")

# import pandas as pd

# # Load the dataset
# df = pd.read_csv('Resources/client_dataset.csv')

# # Ensure data types are correct
# df['unit_price'] = df['unit_price'].astype(float)
# df['qty'] = df['qty'].astype(int)
# df['unit_weight'] = df['unit_weight'].astype(float)

# # Calculate subtotal for each order line
# df['subtotal'] = df['unit_price'] * df['qty']
# print("Subtotal sum:", df['subtotal'].sum())

# # Calculate total weight for each order line
# df['total_weight'] = df['unit_weight'] * df['qty']
# print("Total weight sum:", df['total_weight'].sum())

# # Define the shipping price based on the total weight
# df['shipping_price'] = df['total_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
# print("Shipping price sum:", df['shipping_price'].sum())

# # Calculate the sales tax
# df['sales_tax'] = df['subtotal'] * 0.0925
# print("Sales tax sum:", df['sales_tax'].sum())

# # Calculate the total price
# df['total_price'] = df['subtotal'] + df['shipping_price'] + df['sales_tax']
# print("Total price sum:", df['total_price'].sum())

# # Check if order ID 2173913 had a total price of $162,388.71
# order_id_to_check = 2173913
# expected_total_price = 162388.71

# # Verify the total price for the specific order ID
# order = df[df['order_id'] == order_id_to_check]

# if order.empty:
#     print(f"Order ID {order_id_to_check} not found.")
# else:
#     # Display intermediate calculations for the specific order
#     print(order[['unit_price', 'qty', 'unit_weight', 'subtotal', 'total_weight', 'shipping_price', 'sales_tax']])
    
#     actual_subtotal = order['subtotal'].sum()
#     actual_total_weight = order['total_weight'].sum()
#     actual_shipping_price = order['shipping_price'].sum()
#     actual_sales_tax = order['sales_tax'].sum()
#     actual_total_price = order['total_price'].sum()

#     print(f"Order ID {order_id_to_check} intermediate values:")
#     print(f"Subtotal: {actual_subtotal}")
#     print(f"Total Weight: {actual_total_weight}")
#     print(f"Shipping Price: {actual_shipping_price}")
#     print(f"Sales Tax: {actual_sales_tax}")
#     print(f"Actual Total Price: {actual_total_price}")

#     # Compare the actual total price with the expected total price
#     if abs(actual_total_price - expected_total_price) < 1e-2:  # Allowing for floating-point precision issues
#         print(f"Order ID {order_id_to_check} has the expected total price of ${expected_total_price}")
#     else:
#         print(f"Order ID {order_id_to_check} does not match the expected total price. Actual: ${actual_total_price}, Expected: ${expected_total_price}")


this is the subtotal sum 109440.76
this is the total_weight sum 5576.57
this is the shipping_price 39198.799999999996
this is the sales_tax sum 10123.2703
158762.83029999997
         total_price  subtotal  total_weight
627      9736.298250   7718.90        186.20
14089     503.402400    118.08         37.44
14938    7031.978025   5835.33         93.84
14962    4844.734650   2145.78        357.21
22004    9624.510225   8655.57         16.83
27713    4179.679375   2660.75        181.83
53155  122842.227375  82306.35       4703.22
Order ID 2173913 actual total price: 158762.83029999997
Order ID 2173913 does not match the expected total price. Actual: $158762.83029999997, Expected: $162388.71


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [25]:

# Identify the top 5 clients by total quantity of items ordered
top_clients = df.groupby('client_id')['qty'].sum().nlargest(5).index
#print("top_clients per order line:", df['top_clients'])

# Calculate the total amount spent by each of the top 5 clients
top_clients_spending = df[df['client_id'].isin(top_clients)].groupby('client_id')['total_price'].sum()
#print("top_clients_spending per order line:", df['top_clients_spending'])

# Display the result
print("Total amount spent by each of the top 5 clients by quantity:")
print(top_clients_spending)

Total amount spent by each of the top 5 clients by quantity:
client_id
30143    3.249802e+07
77354    6.973979e+07
78965    6.554297e+08
85931    2.364807e+08
86721    1.753152e+08
Name: total_price, dtype: float64


In [27]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Identify the top 5 clients by total quantity of items ordered
top_clients = df.groupby('client_id')['qty'].sum().nlargest(5).index

# Calculate the total amount spent by each of the top 5 clients
top_clients_data = df[df['client_id'].isin(top_clients)]

# Summarize data for top 5 clients
summary_df = top_clients_data.groupby('client_id').agg(
    total_units_purchased=pd.NamedAgg(column='qty', aggfunc='sum'),
    total_shipping_price=pd.NamedAgg(column='shipping_price', aggfunc='sum'),
    total_revenue=pd.NamedAgg(column='total_price', aggfunc='sum'),
    total_profit=pd.NamedAgg(column='profit', aggfunc='sum')
).reset_index()

# Display the summary DataFrame
print("Summary DataFrame for the top 5 clients by quantity:")
print(summary_df)

Summary DataFrame for the top 5 clients by quantity:
   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      30143                 846069            8981905.24   3.249802e+07   
1      77354                 506382           23814685.86   6.973979e+07   
2      78965                4033302            5340756.88   6.554297e+08   
3      85931                 644423           10122549.19   2.364807e+08   
4      86721                1010426           64335727.72   1.753152e+08   

   total_profit  
0  1.097297e+07  
1  2.770308e+07  
2  6.038261e+07  
3  2.928788e+07  
4  7.373216e+07  


In [28]:
# Format the data and rename the columns to names suitable for presentation.
# Define the money columns. 
# Define a function that converts a dollar amount to millions.
# Apply the currency_format_millions function to only the money columns. 
# Rename the columns to reflect the change in the money format. 
# Summarize data for top 5 clients
summary_df = top_clients_data.groupby('client_id').agg(
    total_units_purchased=pd.NamedAgg(column='qty', aggfunc='sum'),
    total_shipping_price=pd.NamedAgg(column='shipping_price', aggfunc='sum'),
    total_revenue=pd.NamedAgg(column='total_price', aggfunc='sum'),
    total_profit=pd.NamedAgg(column='profit', aggfunc='sum')
).reset_index()

# Define the money columns
money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions
def currency_format_millions(amount):
    return f"${amount / 1_000_000:.2f}M"

# Apply the currency format function to only the money columns
for column in money_columns:
    summary_df[column] = summary_df[column].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format
summary_df.rename(columns={
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price (in millions)',
    'total_revenue': 'Total Revenue (in millions)',
    'total_profit': 'Total Profit (in millions)'
}, inplace=True)

# Display the summary DataFrame
print("Summary DataFrame for the top 5 clients by quantity:")
print(summary_df)


Summary DataFrame for the top 5 clients by quantity:
   client_id  Total Units Purchased Total Shipping Price (in millions)  \
0      30143                 846069                             $8.98M   
1      77354                 506382                            $23.81M   
2      78965                4033302                             $5.34M   
3      85931                 644423                            $10.12M   
4      86721                1010426                            $64.34M   

  Total Revenue (in millions) Total Profit (in millions)  
0                     $32.50M                    $10.97M  
1                     $69.74M                    $27.70M  
2                    $655.43M                    $60.38M  
3                    $236.48M                    $29.29M  
4                    $175.32M                    $73.73M  


In [30]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

# Extract numeric values from the formatted profit column for sorting
summary_df['Total Profit (in millions)'] = summary_df['Total Profit (in millions)'].replace({'\$': '', 'M': ''}, regex=True).astype(float)

# Sort the updated DataFrame by "Total Profit (in millions)" from highest to lowest
sorted_summary_df = summary_df.sort_values(by='Total Profit (in millions)', ascending=False)

# Reformat the "Total Profit (in millions)" column back to the currency format
sorted_summary_df['Total Profit (in millions)'] = sorted_summary_df['Total Profit (in millions)'].apply(lambda x: f"${x:.2f}M")

# Display the sorted summary DataFrame
print("Sorted Summary DataFrame by Total Profit (in millions) from highest to lowest:")
print(sorted_summary_df)

Sorted Summary DataFrame by Total Profit (in millions) from highest to lowest:
   client_id  Total Units Purchased Total Shipping Price (in millions)  \
4      86721                1010426                            $64.34M   
2      78965                4033302                             $5.34M   
3      85931                 644423                            $10.12M   
1      77354                 506382                            $23.81M   
0      30143                 846069                             $8.98M   

  Total Revenue (in millions) Total Profit (in millions)  
4                    $175.32M                    $73.73M  
2                    $655.43M                    $60.38M  
3                    $236.48M                    $29.29M  
1                     $69.74M                    $27.70M  
0                     $32.50M                    $10.97M  
